### ISSUE TITLE
EKS Optimized AMI - Userdata not executed?

### ISSUE THREAD

**Issue by robwittman** [terraform-aws-modules/terraform-aws-eks#294]
# I have issues
getting userdata to run correctly when applied to worker_group
## I'm submitting a...

* [ ] bug report
* [ ] feature request
* [x] support request
* [x] kudos, thank you, warm fuzzy

## What is the current behavior?
We want to do some additional configuration when instances launch in the autoscaling group  to authenticate with private registry. The userdata shows on the instance, `/var/log/cloud-init-output.log` shows the script ran without errors, but none of the files exist when logging in through SSM, and private registry authentication is still failing (as well as other configurations we were doing). 

The configuration is showing that it exists on the EC2 instance 
`curl http://169.254.169.254/latest/user-data`
```
#!/bin/bash -xe

# Allow user supplied pre userdata code
mkdir /docker
echo '{
  "auths": {
    ...redacted...
  }
}
' >> /docker/config.json
echo "DOCKER_CONFIG=/docker" >> /etc/environment


# Bootstrap and join the cluster
/etc/eks/bootstrap.sh --b64-cluster-ca '<redacted>' --apiserver-endpoint '<redacted>' --kubelet-extra-args '' 'cluster-name'

# Allow user supplied userdata code
```
and `/var/log/cloud-init-output.log`
```
Complete!
Cloud-init v. 18.2-72.amzn2.0.6 running 'modules:final' at Tue, 26 Feb 2019 17:22:23 +0000. Up 22.05 seconds.
+ mkdir /docker
+ echo '{
  "auths": {
    ...redacted...
  }
}
'
+ echo DOCKER_CONFIG=/docker
+ /etc/eks/bootstrap.sh <redacted_configuration>
Cluster "kubernetes" set.
Created symlink from /etc/systemd/system/multi-user.target.wants/kubelet.service to /etc/systemd/system/kubelet.service.
```

## If this is a bug, how to reproduce? Please include a code sample if relevant.

`userdata.sh.tpl`
```
mkdir /docker
echo '${dockerconfig}' >> /docker/config.json
echo "DOCKER_CONFIG=/docker" >> /etc/environment
```

`module.tf`
```

data "template_file" "dockerconfig" {
  template = "${file("${path.module}/templates/dockerconfig.json")}"

  vars {
    docker_email      = "${var.docker_email}"
    docker_auth_token = "${var.docker_auth_token}"
  }
}

data "template_file" "userdata" {
  template = "${file("${path.module}/templates/userdata.sh.tpl")}"

  vars {
    dockerconfig = "${data.template_file.dockerconfig.rendered}"
  }
}


locals {
  worker_groups = [
    {
      ...
      ami_id               = "${var.ami_id}"
      additional_userdata  = "${var.worker_userdata}"
      pre_userdata         = "${data.template_file.userdata.rendered}"
    },
  ]
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "2.2.1"
...
  worker_groups                        = "${local.worker_groups}"
  worker_group_count                   = "1"
...
}

```

## What's the expected behavior?
The expected behavior would be that the userdata runs as shown, and creates the desired directories, and performs other actions as instructed. I'm not 100% sure if this would be an issue with this module or how we're using it, or if the problem exists in the EKS Optimized AMI. But I would appreciate any help you guys can give!

## Environment details

* Affected module version: 2.2.1
* Terraform version: 11.10

## Any other relevant info
AMI ID: `ami-0c28139856aaf9c3b` (Optimized EKS in us-west-2)
Connecting to instance through SSM

**Comment by robwittman** [comment]
I did some more debugging, and it doesn't seem to be an issue with this module at all. If we add an SSH key to the worker nodes and access them directly as `ec2-user`, it looks like all the expected commands are executed, and the desired files and updates exist. However, if we access the same node through SSM, we're unable to see any of those changes.

We're currently debugging the SSM issue, and believe it's either directly related to SSM, or our usage of the `ssm-agent` DaemonSet running in the cluster. 

I'll update again when we find out what the root cause was, in case any one else experiences it.

**Comment by brandoconnor** [comment]
Thanks for digging in a bit further @robwittman . I'll close this down for now but feel free to comment back if you have further findings. Always good to know where the edge cases are and work to steer folks away from them however possible.

**Comment by KYannick** [comment]
@robwittman Did you eventually find the SSM issue?

**Comment by robwittman** [comment]
@KYannick After we launched `ssm-agent` directly on the node(s), and accessed them that way, we were able to see that instance was running the userdata as expected. 

It seems the issue is with either with the `kube-ssm-agent`, or our configuration of it. When accessing the node through the `kube-ssm-agent` daemonset, I get a completely different 'host' environment. I haven't done much more looking into the `kube-ssm-agent` side, since bootstrapping SSM directly on the node worked as expected. 

We just use
```
# userdata.sh
sudo yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm
```
directly from AWS documentation

**Comment by mvillumsen** [comment]
@robwittman Thanks for creating this issue. It helped me solve a similar issue (pulling from private registry). I ended up using the solution you describe in the issue. With some help from https://kubernetes.io/docs/concepts/containers/images/#configuring-nodes-to-authenticate-to-a-private-registry I figured out that the docker config should be stored in `/var/lib/kubelet/config.json` for it the work but besides that your solution works just fine. Thanks :)
